### PR TITLE
add Log 

### PR DESCRIPTION
**Pull request by xfc1939**

### CODE REVIEW

File: src/common/Log/Log.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
+//
+// Created by XFC on 2018/6/24.
+//
+
+#include "Log.h"
+
+char logLevelStr[][8] = {"[DEBUG]", "[TRACE]", "[WARN]", "[ERROR]"};
```
Comment (wangqing1218) [inline src/common/Log/Log.cpp]: `logLevelStr`'s type is not make sense.  `char [][8]` is so magical. 
Maybe string[] will be better.

File: src/common/Log/Log.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+LogContext* LogContext::instance() {
+    static LogContext _context;
```
Comment (wangqing1218) [inline src/common/Log/Log.cpp]: `_` should not be the prefix in local variable.

File: src/common/Log/Log.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    return &_context;
```
Comment (wangqing1218) [inline src/common/Log/Log.cpp]: returning a ref to instance will be better.

File: src/common/Log/Log.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+void LogContext::setCacheSize(int size) {
+    m_maxCacheSize = size;
+    m_cache.resize(size);
+}
+
+void LogContext::addHandle(const std::function<void(const std::string &msg)> &func, bool enableCache){
```
Comment (wangqing1218) [inline src/common/Log/Log.cpp]: addHandler

Comment (wangqing1218) [inline src/common/Log/Log.cpp]: `func` is ambiguous. it should be hander

File: src/common/Log/Log.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+#define SNAIL_LOGCONTEXT_H
+
+#include <list>
+#include <functional>
+#include <sstream>
+#include <vector>
+#include <map>
+#include <thread>
+
+enum Loglevel {
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: LogLevel should under the space of Log.

File: src/common/Log/Log.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    Debug = 0,
+    Trace,
+    Warn,
+    Error
+};
+
+class LogContext {
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: Why is `LogContext` a singleton class?

File: src/common/Log/Log.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+public:
+    static LogContext* instance();
+public:
+    void addHandle(const std::function<void(const std::string &msg)> &func, bool enableCache);
+    void flush(const std::string &msg);
+    void setCacheSize(int size);
+
+private:
+    LogContext():m_cache(), m_currCacheSize(0), m_maxCacheSize(0){}
+    ~LogContext() = default;
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: virtual dtor here will be better.

File: src/common/Log/Log.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+private:
+    std::list<std::function<void(const std::string &msg)>> m_noCacheFunc;
+    std::list<std::function<void(const std::string &msg)>> m_cacheFunc;
+    std::string m_cache;
+    int m_maxCacheSize;
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: For a private member variable, it should be named as which postfix is `_`, like `maxCacheSize_`.
and public like `maxCacheSize`.

File: src/common/Log/Log.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    void addHandle(const std::function<void(const std::string &msg)> &func, bool enableCache);
+    void flush(const std::string &msg);
+    void setCacheSize(int size);
+
+private:
+    LogContext():m_cache(), m_currCacheSize(0), m_maxCacheSize(0){}
+    ~LogContext() = default;
+
+private:
+    std::list<std::function<void(const std::string &msg)>> m_noCacheFunc;
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: The type is too log to read, and we can cut it by `using`.
```c++
using HandlerListType = xxx;
```

File: src/common/Log/Log.h [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+ public:
+     Log() = default;
+     ~Log();
+
+ public:
+    Log& operator << (Loglevel level);
+     Log& operator << (int value);
+     Log& operator << (const char *pChar);
+ private:
+     static Loglevel m_logLevel;
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: I don't know  why does `m_logLevel ` is a private static variable?

File: src/common/Log/Log.h [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+ private:
+     std::ostringstream m_oStream;
+     Loglevel m_currentLogLevel;
+ };
+
+#define Debug() (Log() << Loglevel::Debug <<__TIME__ << __FILE__ << __LINE__)
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: Where are Info, Error, etc..

File: src/common/Log/LogTest.cpp [hunk 12]
```diff
@@ -0,0 +1,27 @@
+//
```
Comment (wangqing1218) [inline src/common/Log/LogTest.cpp]: We should create a main file to test it.
In general, we must write unit tests and integration tests to make sure that the code is make sense.
But, we have not any test tool now. we should add google test to it firstly.

File: src/common/Log/Log.cpp [hunk 13]
```diff
@@ -0,0 +1,79 @@
+//
```
Comment (wangqing1218) [inline src/common/Log/Log.cpp:1]: All these codes should be under namespace snail.

File: src/common/Log/Log.cpp [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+                    func(m_cache_ + msg);
+                }
+                m_currCacheSize_ = 0;
+                m_cache_.clear();
+            } else {
+                m_cache_.append(msg);
+            }
+        }
+
+        Loglevel Log::m_logLevel = Loglevel::Debug;
```
Comment (wangqing1218) [inline src/common/Log/Log.cpp]: Why this here?

File: src/common/Log/Log.cpp [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        LogContext* Log::m_context = nullptr;
+        const std::vector<std::string> Log::LogLevelStr   = {"[DEBUG]", "[TRACE]", "[WARN]", "[ERROR]"};
+
+        Log::~Log() {
+            if (m_currentLogLevel_ >= m_logLevel) {
```
Comment (wangqing1218) [inline src/common/Log/Log.cpp]: why flush here?

File: src/common/Log/Log.h [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+            static Loglevel m_logLevel;
+            static LogContext *m_context;
+
+        private:
+            std::ostringstream m_oStream_;
+            Loglevel m_currentLogLevel_;
+        };
+    }
+}
+#define Debug() (snail::log::Log() << snail::log::Loglevel::Debug <<__TIME__ << __FILE__ << __LINE__)
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: I don't think the format is ok.

File: src/test/LogTest.cpp [hunk 17]
```diff
@@ -0,0 +1,27 @@
+//
+// Created by XFC on 2018/6/24.
+//
+#include <iostream>
+#include <fstream>
+#include <thread>
+
+#include "../common/Log/Log.h"
```
Comment (wangqing1218) [inline src/test/LogTest.cpp]: We should add the path into header paths in cmake. 
```
#include "Log/Log.h"
```

File: src/test/LogTest.cpp [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+int main() {
+    snail::log::LogContext context;
```
Comment (wangqing1218) [inline src/test/LogTest.cpp]: We needn't the namespace `Log`.
`snail` is enough.

File: src/test/LogTest.cpp [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    snail::log::Log::setLogContext(&context);
+
+    snail::log::Log::addHandle([](const std::string &message) {
+        std::cout << message << std::endl;
+    }, false);
+    snail::log::Log::addHandle([](const std::string &message){
```
Comment (wangqing1218) [inline src/test/LogTest.cpp]: addHander

File: src/common/Log/Log.h [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+        public:
+            Log &operator<<(Loglevel level);
+
+            Log &operator<<(int value);
+
+            Log &operator<<(const char *pChar);
+
+        private:
+            static Loglevel m_logLevel;
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: why static variable is what's prefix is m?

File: src/common/Log/Log.h [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+            std::list<LogHandler> m_cacheFunc_;
+            std::string m_cache_;
+            int m_maxCacheSize_;
+            int m_currCacheSize_;
+        };
+
+        class Log {
+        public:
+            static void setLoglevel(Loglevel level);
+            static void addHandle(const LogHandler &handler, bool enableCache);
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: addHandler

File: src/common/Log/Log.h [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+            LogContext() : m_cache_(), m_currCacheSize_(0), m_maxCacheSize_(0) {}
+            virtual ~LogContext() = default;
+
+        public:
+            void addHandler(const LogHandler &handler, bool enableCache);
+            void flush(const std::string &msg);
+            void setCacheSize(int size);
+
+        private:
+            std::list<LogHandler> m_noCacheFunc_;
```
Comment (wangqing1218) [inline src/common/Log/Log.h]: noCacheFuncs

**Review by wangqing1218 — CHANGES_REQUESTED** [review]

File: src/common/Log/Log.h (hunk 16 above)
Comment (xfc1939) [reply src/common/Log/Log.h]: what do you think?

File: src/test/LogTest.cpp (hunk 19 above)
Comment (xfc1939) [reply src/test/LogTest.cpp]: Are you sure? hander => 投手，handle => 句柄

File: src/common/Log/Log.h (hunk 21 above)
Comment (xfc1939) [reply src/common/Log/Log.h]: I get this result  "Handler => 处理器  handle => 句柄" from Google translate website

File: src/common/Log/Log.h (hunk 22 above)
Comment (xfc1939) [reply src/common/Log/Log.h]: ok, I  will reslove this spell error.

File: src/test/LogTest.cpp (hunk 18 above)
Comment (xfc1939) [reply src/test/LogTest.cpp]: ok

File: src/common/Log/Log.cpp (hunk 15 above)
Comment (xfc1939) [reply src/common/Log/Log.cpp]: otherwise , do yo think where is  this func should be?

Comment (wangqing1218) [reply src/common/Log/Log.cpp]: I don't know yet.

File: src/common/Log/Log.h (hunk 16 above)
Comment (wangqing1218) [reply src/common/Log/Log.h]: OK. Nothing.  Why Debug under the name Loglevel?

File: src/test/LogTest.cpp (hunk 19 above)
Comment (wangqing1218) [reply src/test/LogTest.cpp]: Handler is the noun of handle.
handler here means 要处理的东西。
Anyway, 句柄 here is not suitable.

File: src/common/Log/Log.h (hunk 22 above)
Comment (wangqing1218) [reply src/common/Log/Log.h]: Not spelling lint error. I means that for a member, we should write it as a `m` prefix.

File: src/test/LogTest.cpp (hunk 19 above)
Comment (xfc1939) [reply src/test/LogTest.cpp]: handle is a noun too

Comment (wangqing1218) [reply src/test/LogTest.cpp]: 句柄 here is not suitable. `addhandle`  means 添加一个句柄？
but your parameter is not a handle.

**Comment by wangqing1218** [comment]
Where is the unit test?

**Review by wangqing1218 — APPROVED** [review]
LGTM

**Pull request merged**